pymc_marketing/clv/distributions.py

```diff
@@ -483,3 +480,184 @@ def logp(value, r, alpha, s, beta, T):
+
+        return output
+
+    def _supp_shape_from_params(*args, **kwargs):
+        return (2,)
+
+
+beta_geo_beta_binom = BetaGeoBetaBinomRV()
+
+
+# TODO: Edit likelihood expression in docstring
+class BetaGeoBetaBinom(PositiveContinuous):
+    r"""
+    Population-level distribution class for a discrete, non-contractual, Beta-Geometric/Beta-Binomial process,
+    based on Fader, et al. in [1]_.
+
+    .. math::
+
+        \begin{align}
+        \text{if }\alpha > \beta: \\
+        \\
+        \mathbb{L}(r, \alpha, s, \beta | x, t_x, T) &=
+        \frac{\Gamma(r+x)\alpha^r\beta}{\Gamma(r)+(\alpha +t_x)^{r+s+x}}
+        [(\frac{s}{r+s+x})_2F_1(r+s+x,s+1;r+s+x+1;\frac{\alpha-\beta}{\alpha+t_x}) \\
+        &+ (\frac{r+x}{r+s+x})
+        \frac{_2F_1(r+s+x,s;r+s+x+1;\frac{\alpha-\beta}{\alpha+T})(\alpha +t_x)^{r+s+x}}
+        {(\alpha +T)^{r+s+x}}] \\
+        \\
+        \text{if }\beta >= \alpha: \\
+        \\
+        \mathbb{L}(r, \alpha, s, \beta | x, t_x, T) &=
+        \frac{\Gamma(r+x)\alpha^r\beta}{\Gamma(r)+(\beta +t_x)^{r+s+x}}
+        [(\frac{s}{r+s+x})_2F_1(r+s+x,r+x;r+s+x+1;\frac{\beta-\alpha}{\beta+t_x}) \\
+        &+ (\frac{r+x}{r+s+x})
+        \frac{_2F_1(r+s+x,r+x+1;r+s+x+1;\frac{\beta-\alpha}{\beta+T})(\beta +t_x)^{r+s+x}}
+        {(\beta +T)^{r+s+x}}]
+        \end{align}
+
+    ========  ===============================================
+    Support   :math:`t_j > 0` for :math:`j = 1, \dots, x`
+    Mean      :math:`\mathbb{E}[X(t) | r, \alpha, s, \beta] = \frac{r\beta}{\alpha(s-1)}[1-(\frac{\beta}{\beta + t})^{s-1}]`
+    ========  ===============================================
+
+    References
+    ----------
+    .. [1] Fader, Peter S., Bruce G.S. Hardie, and Jen Shang (2010),
+       "Customer-Base Analysis in a Discrete-Time Noncontractual Setting,"
+       Marketing Science, 29 (6), 1086-1108. https://www.brucehardie.com/papers/020/fader_et_al_mksc_10.pdf
+
+    """
+    rv_op = beta_geo_beta_binom
+
+    @classmethod
+    def dist(cls, alpha, beta, gamma, delta, T, **kwargs):
+        return super().dist([alpha, beta, gamma, delta, T], **kwargs)
+
+    def logp(value, alpha, beta, gamma, delta, T):
+        t_x = value[..., 0]
+        x = value[..., 1]
+
+        betaln_ab = betaln(alpha, beta)
+        betaln_gd = betaln(gamma, delta)
+
+        A = (
+            betaln(alpha + x, beta + T - x)
+            - betaln_ab
+            + betaln(gamma, delta + T)
+            - betaln_gd
+        )
+
+        # TODO: The -1 may need to be omitted for the sequences loop
+        t_recent = T - t_x - 1
+
+        # TODO: Resolve this scan loop and logp will be ready
+        def _B_iter(ix):
+            return betaln(alpha + x, beta + t_x - x + ix) + betaln(
+                gamma + 1, delta + t_x + ix
+            )
+
+        iter_result, _ = scan(
+            fn=_B_iter,
+            # outputs_info=pt.zeros(1),
+            sequences=pt.arange(t_recent),
+        )
```

Comment:
If I understand what you're trying to do, it should be something like:

```suggestion
        def _B_iter(ix, alpha, beta, t_x, x, gamma, delta):
            return betaln(alpha + x, beta + t_x - x + ix) + betaln(
                gamma + 1, delta + t_x + ix
            )

        iter_result, _ = scan(
            fn=_B_iter,
            outputs_info=[None],
            sequences=pt.arange(t_recent),
            non_sequences=[alpha, beta, t_x, x, gamma, delta],
        )
```
`outputs_info=[None]` tells Pytensor it expects a single non-recurrent output
`non_sequences` passes the constants explicitly to the inner function, which is a bit more clean imho

Reply:
Note that your scan can be replaced by relying on broadcasting.

```suggestion
        n_dims_expr = max([alpha.ndim, beta.ndim, x.ndim, t_x.ndim, gamma.ndim, delta.ndim, t_recent.ndim])
        extra_dims_needed = (n_dims_expr - t_recent.ndim) + 1
        ixs = pt.shape_padright(pt.arange(t_recent), extra_dims_needed)
        iter_result = (
          betaln(alpha + x, beta + t_x - x + ixs)
          + betaln(gamma + 1, delta + t_x + ixs)
        )
```